crates/publish/src/error.rs

```diff
@@ -0,0 +1,40 @@
+/// A specialized Result type for publish operations
+pub type PublishResult<T> = std::result::Result<T, Error>;
+
+/// Describes various errors that can be returned during publishing
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// Failed because the path terminates in a root or prefix
+    #[error("Failed to get containing directory for app file '{0}'")]
+    AppDir(String),
+    /// The bindle already exists
+    #[error("Bindle {0} already exists on the server")]
+    BindleAlreadyExists(String),
+    /// Bindle client failure
+    #[error("Error creating bindle client")]
+    BindleClient(#[from] bindle::client::ClientError),
+    /// Malformed bindle id
+    #[error("App name and version '{0}' do not form a bindle ID")]
+    BindleId(String),
+    /// IO error happened during building bindle component manifest
+    #[error("Failed to build bindle component manifest: {path}")]
+    BindleComponentManifest {
+        /// Represents path on a filesystem
+        path: String,
```

Comment:
Given your purpose in this PR, I wonder if it's worth distinguishing "I couldn't find the `component.source` file" (Wasm file not built) from "something went wrong getting asset file details" (possibly a build issue, but possibly more generic I/O problem)?